### PR TITLE
refactor SourceInfo trait to remove buffer_message 

### PR DESCRIPTION
**Pull request by ruchirK**
This pr tries to refactor the operator returned by `create_source` (or I guess, more precisely, the inner closure returned by that function) to no longer rely on each specific `SourceInfo` implementation to handle message buffering.

I believe that we should reduce the amount of functionality in `SourceInfo` to only cover "reading from upstream sources, and reporting how far we have read" instead of the many different things that trait does currently.

The first commit is purely code movement to extract out the core bit of logic required to handle ingesting a message (the resulting function is a bit of a mess but baby steps) and then the second commit actually removes the per-source info buffering

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6053)

### CODE REVIEW

**Comment by ruchirK** [comment]
I think file / s3 sources are totally fine with this change, because now the source operator is doing the same exact buffering that the FileSourceInfo or S3SourceInfo would have done otherwise. Kafka and Kinesis sources I'm less sure of so I'm going to spend some time trying to suss that out now.

**Comment by ruchirK** [comment]
Ok so as best I can tell this only changes behavior for the kafka case. In kafka sources, when a message is buffered at a partition, the next time we try to read from the source we go to a different partition (if there happens to be multiple partitions assigned to the same worker). Its unclear to me what value the current behavior gives us (because this would only be relevant under the BYO timestamping world + in a scenario where more than one partition is assigned to the same worker)

I guess best case it means we can still make a marginal amount of read progress if we haven't received the next timestamping information for a partition. No matter how I look at it this seems not critical for correctness

File: src/dataflow/src/source/kafka.rs [hunk 1]
```diff
@@ -121,44 +116,14 @@ impl SourceInfo<Vec<u8>> for KafkaSourceInfo {
...
-        // Case 2.a
-        {
+        // For transactional or compacted topics, the `last_offset` may not correspond to
+        // an accurate representation of how far we have read up to (because subsequent offsets
+        // have been garbage collected). However, `last_offset` still represents the lower bound
+        // for the value of the next offset in this partition.
+        if !self.has_partition(kafka_pid) || last_offset >= offset {
             true
         } else {
-            let mut current_consumer_position: MzOffset = KafkaOffset {
```
Comment (ruchirK) [inline src/dataflow/src/source/kafka.rs:139]: I'm not 100% sure about this block of changes otherwise I feel good about everything else in this pr

Comment (ruchirK) [reply src/dataflow/src/source/kafka.rs:139]: my understanding is that the previous code to check the current kafka read position was a performance optimization and not required for correctness but I will try to validate that more carefully

Comment (ruchirK) [reply src/dataflow/src/source/kafka.rs:139]: the current logic was introduced in https://github.com/MaterializeInc/materialize/pull/3259

Comment (ruchirK) [reply src/dataflow/src/source/kafka.rs:139]: @elindsey let me know if you have thoughts here

Comment (ruchirK) [reply src/dataflow/src/source/kafka.rs:139]: Ok I did some slack sleuthing I think the original motivation for the "check current position" logic came from https://github.com/MaterializeInc/materialize/issues/3265 but I'm not sure if thats something to be concerned about anymore

Comment (elindsey) [reply src/dataflow/src/source/kafka.rs:139]: I'm pretty sure none of this is applicable with rt. rt in its current form doesn't care about offsets at all

**Review by elindsey — APPROVED** [review]
lgtm! I'm happy to see handle_message moved to its own fn - step in the right direction towards sorting out source/mod.rs. I found this area of the code very difficult to change when I started

**Comment by ruchirK** [comment]
Doubled checked just to be safe -- both this and main have the same behavior on empty topics.

**Pull request merged**